WLC/code_fixing/codefixer.py

```diff
@@ -1,92 +1,181 @@
+        if prev_word == "import" or prev_word == "for" or next_word == "=":
+            if word and word not in self.context:
+                self.context.append(word)  # add the imported module, or the var name to contextual data
+            return word
+
+        elif word.startswith('--') and word.endswith('--'):
+            no_dash = word.strip('--')
+            best_word, start, end = self._levenshtein(no_dash, check_on=SPEC_ATTR)
+            if best_word and len(best_word) == len(no_dash):
+                word = "__" + best_word + "__"
+            else:
+                word = "__" + no_dash + "__"
+
+        elif not prev_word:
+            best_word, start, end = self._levenshtein(word)
+            if best_word:
+                word = word[:start] + best_word + word[end:]
+
+        return word
+
+    def _fix_line(self, line, line_number):
         fixed_words = []
-        for word in line.split():
-            prev_word = fixed_words[-1] if fixed_words else None
-
-            prev_context = prev_word if prev_word in ["def", "class", "=", "import"] else False
-            new_word, contextual_data = self._fix_word(word, contextual_data, prev_context)
-
-            if prev_word == "class" or prev_word == "def":
-
-                if "(" in new_word:
-                    name = new_word[:new_word.find("(")]
-                    contextual_data.append(name)
-
-                if "(" in new_word and ")" in new_word:
-                    args = new_word[new_word.find("(") + 1:new_word.rfind(")")]
-                    # TODO: contextual data check on args
-
-            if prev_word == "=" and len(fixed_words) > 2:
-                    contextual_data.append(fixed_words[-2])  # add var name before = sign to contextual data
-
-            fixed_words.append(new_word)
-
-        new_line = " ".join(fixed_words)
-
-        # If any of these, I expect it to end with a colon
-        # NOTE: this is currently hardcoding; not interesting.
-        # if any(new_line.startswith(b) for b in ["class", "def", "if", "for"]) and new_line.endswith("i"):
-        #     new_line = new_line[:-1] + ":"
-
-        return new_line, contextual_data
+        words = line.split()
+        for word_number, word in enumerate(words):
+            prev_word = fixed_words[-1] if fixed_words and fixed_words[-1] in ["def", "class", "import"] else None
+            next_word = words[word_number + 1] if len(words) > word_number + 1 else None
+
+            fixed_word = word  # set default
+
+            if prev_word == "class" or prev_word == "def":  # class/method/function declaration
+                self._set_decl(word)
+
+            elif prev_word == "for":  # se var X as context in a 'for X in _' statement
+                self._set_forloop_var(word)
+
+            elif "(" in word and ")" in word:  # method/function call
+                fixed_word = self._fix_funccall(word, line_number, word_number)
```

Comment:
doesn't this also visit math expressions?

Reply:
You're right, it does - I'll make sure it's prepended with a piece of contextual text.